lib/ecto/parameterized_type.ex

```diff
@@ -0,0 +1,83 @@
+defmodule Ecto.ParameterizedType do
+  @moduledoc """
+  Parameterized types are similar to Ecto Custom Types, however they allow a set of options to be
+  specified in the schema which are initialized on compilation and passed to the callback functions
+  as a final argument. Note that options should be specified as a keyword, but it is idiomatic to
+  convert them to maps inside init for easier pattern matching in other callbacks.
+
+  ## Examples
+
+    To create a parameterized type, create a module as shown below:
+
+    defmodule MyApp.MyType do
+      use Ecto.ParameterizedType
+
+      def type(_opts), do: :string
+
+      def init(opts) do
+        ...
+        Enum.into(opts, %{})
+      end
+
+      def cast(data, current, opts) do
+        ...
+        cast_data
+      end
+
+      def load(data, opts) do
+        ...
+        {:ok, loaded_data}
+      end
+
+      def dump(data, opts) do
+        ...
+        {:ok, dumped_data}
+      end
+
+      def equal?(a, b, _opts) do
+        a == b
+      end
+    end
+
+  To use this type in a schema field, specify it in the form {module, keyword_opts} like this:
+
+    schema "foo" do
+      field "bar", {MyApp.MyType, opt1: :baz, opt2: :boo}
+    end
+
+  """
+
+  @type opts :: keyword() | map()
+
+  @callback init(opts :: keyword()) :: opts()
+
+  @callback cast(data :: term, current :: term, opts :: opts()) ::
+              {:ok, term} | {:error, keyword()} | :error
+
+  @callback load(value :: any(), opts :: opts()) :: {:ok, value :: any()} | :error
+
+  @callback load(value :: any(), loader :: function(), opts :: opts()) :: {:ok, value :: any()} | :error
+
+  @callback dump(value :: any(), opts :: opts()) :: {:ok, value :: any()} | :error
+
+  @callback dump(value :: any(), dumper :: function(), opts :: opts()) :: {:ok, value :: any()} | :error
+
+  @callback type(opts :: opts()) :: Ecto.Type.t()
+
+  @callback equal?(value :: any(), value :: any(), opts :: opts()) :: boolean()
+
+  @callback embed_as(format :: atom(), opts :: opts()) :: :self | :dump
+
+  @doc false
+  defmacro __using__(_) do
+    quote location: :keep do
+      @behaviour Ecto.ParameterizedType
+      def load(term, _loader, opts), do: load(term, opts)
+      def dump(term, _dumper, opts), do: dump(term, opts)
```

Comment:
We definitely don't need both `load/2` and `load/3`, so my suggestion is to keep it simply as `load/3` and `dump/3` and simplify the API.